csrc/id_graphs.cpp

```diff
@@ -0,0 +1,3521 @@
+  idGraph(IdMappingMode::ALMOSTEXACT) = idGraph(IdMappingMode::EXACT);
+  idGraph(IdMappingMode::ALMOSTEXACT).mapThroughTrivialExprs();
+  idGraph(IdMappingMode::ALMOSTEXACT).removeTrivialExprs();
+}
+
+void IterDomainGraphs::validateAndPropagatePType() const {
+  for (const auto& loop_disjoint_set :
+       idGraph(IdMappingMode::LOOP).disjointIdSets().disjointSets()) {
+    ParallelType common_ptype = ParallelType::Serial;
+    for (auto id : loop_disjoint_set->vector()) {
+      auto id_ptype = id->getParallelType();
+      TORCH_INTERNAL_ASSERT(
+          id_ptype == common_ptype || id_ptype == ParallelType::Serial ||
+              common_ptype == ParallelType::Serial,
+          "Issue validating parallel type disjoint ptype is, ",
+          common_ptype,
+          " but found in the set the id: ",
+          id->toString());
+      common_ptype =
+          common_ptype == ParallelType::Serial ? id_ptype : common_ptype;
+    }
+
+    for (auto id : loop_disjoint_set->vector()) {
+      id->parallelize(common_ptype);
+    }
+  }
+}
+
+void IterDomainGraphs::build(
+    const std::vector<Expr*>& exprs,
+    const std::vector<TensorView*>& additional_tvs) {
+  // Initialize the required sets as if a permissive relationship is never
+  // found, then querying an empty permissive map will fail later.
+  // Initialize disjoint sets
+  for (auto mode : kIdMappingModes) {
+    id_graphs_[mode] = IdGraph();
+  }
+
+  std::vector<Expr*> tv_exprs;
+
+  std::copy_if(
+      exprs.begin(), exprs.end(), std::back_inserter(tv_exprs), [](Expr* expr) {
+        TORCH_INTERNAL_ASSERT(expr != nullptr);
+        return ir_utils::isTvOp(expr);
+      });
+
+  auto all_tvs = ir_utils::allTvsOfExprs(tv_exprs);
+  if (additional_tvs.size() > 0) {
+    std::unordered_set<TensorView*> all_added_tvs(
+        all_tvs.begin(), all_tvs.end());
+    for (auto additional_tv : additional_tvs) {
+      if (all_added_tvs.find(additional_tv) == all_added_tvs.end()) {
+        all_tvs.push_back(additional_tv);
+      }
+    }
+  }
+
+  if (all_tvs.empty()) {
+    return;
+  }
+
+  FusionGuard fg(all_tvs.front()->fusion());
+  // Add uses and definitions to all iter domains.
+  buildIterDomainDefinitionsAndUses(all_tvs);
+
+  // Initialize the maps with all the IterDomains used in the provded
+  // expressions.
+  idGraph(IdMappingMode::EXACT) = initializeIdGraph();
+
+  std::cout << "buildExactMap" << std::endl;
+  buildExactMap(tv_exprs);
+  std::cout << "buildAlmostExactMap" << std::endl;
+  buildAlmostExactMap();
+  std::cout << "buildPermissiveMap" << std::endl;
+  buildPermissiveMap(tv_exprs);
+  std::cout << "built non lowering graphs" << std::endl;
+
+  // Only build loop map during lowering
+  if (FusionGuard::getCurFusion()->isA<kir::Kernel>()) {
+    FusionGuard::getCurFusion()->print();
+    // Find loops that need to be promoted because of broadcast resolution,
+    // figure out what that resolution should look like, compute IDs for it if
+    // necessary.
```

Comment:
A couple examples might be useful here or better yet in a docstring for the method. I guess mainly I am still not entirely sure what _promotion_ is referring to in this context and why we might need to create new IDs when we resolve a broadcast.